src/coreComponents/physicsSolvers/multiphysics/CompositionalMultiphaseReservoirAndWells.cpp

```diff
@@ -0,0 +1,404 @@
+    ElementRegionManager::ElementViewAccessor< arrayView1d< globalIndex const > > const resDofNumberAccessor =
+      elemManager.constructArrayViewAccessor< globalIndex, 1 >( resDofKey );
+    ElementRegionManager::ElementViewConst< arrayView1d< globalIndex const > > const resDofNumber =
+      resDofNumberAccessor.toNestedViewConst();
+    globalIndex const rankOffset = dofManager.rankOffset();
+
+    elemManager.forElementSubRegions< WellElementSubRegion >( regionNames, [&]( localIndex const,
+                                                                                WellElementSubRegion const & subRegion )
+    {
+
+      // if the well is shut, we neglect reservoir-well flow that may occur despite the zero rate
+      // therefore, we do not want to compute perforation rates and we simply assume they are zero
+      WellControls const & wellControls = Base::getWellSolver()->getWellControls( subRegion );
+      bool const detectCrossflow =
+        ( wellControls.isInjector() ) && wellControls.isCrossflowEnabled() &&
+        getLogLevel() >= 1; // since detect crossflow requires communication, we detect it only if the logLevel is sufficiently high
+
+      PerforationData const * const perforationData = subRegion.getPerforationData();
+
+      // get the degrees of freedom
+      string const wellDofKey = dofManager.getKey( Base::getWellSolver()->wellElementDofName() );
+      arrayView1d< globalIndex const > const & wellElemDofNumber =
+        subRegion.getReference< array1d< globalIndex > >( wellDofKey );
+
+      // get well variables on perforations
+      arrayView2d< real64 const > const & compPerfRate =
+        perforationData->getExtrinsicData< extrinsicMeshData::well::compPerforationRate >();
+      arrayView3d< real64 const > const & dCompPerfRate_dPres =
+        perforationData->getExtrinsicData< extrinsicMeshData::well::dCompPerforationRate_dPres >();
+      arrayView4d< real64 const > const & dCompPerfRate_dComp =
+        perforationData->getExtrinsicData< extrinsicMeshData::well::dCompPerforationRate_dComp >();
+
+      arrayView1d< localIndex const > const & perfWellElemIndex =
+        perforationData->getReference< array1d< localIndex > >( PerforationData::viewKeyStruct::wellElementIndexString() );
+
+      // get the element region, subregion, index
+      arrayView1d< localIndex const > const & resElementRegion =
+        perforationData->getReference< array1d< localIndex > >( PerforationData::viewKeyStruct::reservoirElementRegionString() );
+      arrayView1d< localIndex const > const & resElementSubRegion =
+        perforationData->getReference< array1d< localIndex > >( PerforationData::viewKeyStruct::reservoirElementSubregionString() );
+      arrayView1d< localIndex const > const & resElementIndex =
+        perforationData->getReference< array1d< localIndex > >( PerforationData::viewKeyStruct::reservoirElementIndexString() );
```

Comment:
Still some little `extrinsicMeshData` work!

Reply:
Yes, you are right, I added the `extrinsicMeshData` for these quantities